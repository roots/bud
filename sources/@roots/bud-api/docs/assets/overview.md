---
title: bud.assets
---

Include static assets in your compilation even if they aren't referenced in scripts or stylesheets.
