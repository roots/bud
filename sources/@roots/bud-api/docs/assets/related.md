## Related

- [bud.copyFile](/docs/bud.copyFile)
- [bud.copyDir](/docs/bud.copyDir)
