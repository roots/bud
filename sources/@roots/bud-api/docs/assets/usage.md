---
title: Usage
---

All paths are `@src` relative.

### Copying a directory

Copy the entire `@src/images` directory:

```ts title=bud.config.ts
bud.assets('images')
```

### Copying a file

Copy a single file:

```ts title=bud.config.ts
bud.assets('images/image.png')
```

### Copying from multiple sources

You may add array items to specify additional tasks.

```ts title=bud.config.ts
bud.assets([`images`, `fonts`])
```

Or, call **bud.assets** more than once:

```ts title=bud.config.ts
bud
  .assets(`images`)
  .assets(`fonts`)
```

### Copying using an object

For more granular control, you may specify [`CopyPlugin.ObjectPattern`](https://github.com/webpack-contrib/copy-webpack-plugin#patterns) object(s) directly.

As an example, to copy all the images from `vendor/images` and preserve the directory structure:

```ts title=bud.config.ts
bud.assets({
  from: `vendor/images/**/*`,
  context: bud.path(),
})
```

## Additional information

You don't need to import assets which are utilized by your bundled code. For instance, if you are referencing a font file from your stylesheet, the font will already be included in your distribution. You don't need to manually require it with **bud.assets**, although there is probably no real harm in doing so.

**bud.assets** is specifically for compiling files which are not already included elsewhere.
