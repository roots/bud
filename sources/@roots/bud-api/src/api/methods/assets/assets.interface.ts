import type {Framework} from '@roots/bud-framework'
import CopyPlugin from 'copy-webpack-plugin'

export interface facade {
  (
    ...request: Array<
      | string
      | CopyPlugin.ObjectPattern
      | Array<string | [string, string] | CopyPlugin.ObjectPattern>
    >
  ): Framework
}

export interface method {
  (
    ...request: Array<
      | string
      | CopyPlugin.ObjectPattern
      | Array<string | [string, string] | CopyPlugin.ObjectPattern>
    >
  ): Promise<Framework>
}
