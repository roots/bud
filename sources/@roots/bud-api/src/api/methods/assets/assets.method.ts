import type {Framework} from '@roots/bud-framework'
import {globby} from '@roots/bud-support'
import {dirname} from 'path'

import type {method} from './assets.interface'

export const assets: method = async function assets(
  paths: string[],
): Promise<Framework> {
  this as Framework
  paths = await globby(paths)

  paths.map((src: string) => {
    const dirName = dirname(src).replace(
      this.path('src'),
      this.path('dist'),
    )

    const fileName =
      this.store.is('features.hash', true) && this.isProduction
        ? `${dirName}/${this.store.get('hashFormat')}[ext]`
        : `${dirName}/${this.store.get('fileFormat')}[ext]`

    this.extensions.get('copy-webpack-plugin').mergeOption('patterns', [
      {
        from: src,
        to: fileName,
        noErrorOnMissing: true,
      },
    ])
  })

  return this
}
