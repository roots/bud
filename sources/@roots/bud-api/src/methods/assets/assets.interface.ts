import type {Bud} from '@roots/bud-framework'
import type CopyPlugin from 'copy-webpack-plugin'

export interface facade {
  (
    ...request: Array<
      | string
      | CopyPlugin.ObjectPattern
      | Array<string | [string, string] | CopyPlugin.ObjectPattern>
    >
  ): Bud
}

export interface method {
  (
    ...request: Array<
      | string
      | CopyPlugin.ObjectPattern
      | Array<string | [string, string] | CopyPlugin.ObjectPattern>
    >
  ): Promise<Bud>
}
