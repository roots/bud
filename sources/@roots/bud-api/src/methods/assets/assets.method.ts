import type {Bud} from '@roots/bud-framework'
import CopyPlugin from 'copy-webpack-plugin'
import {isArray, isString} from 'lodash-es'
import {normalize} from 'node:path'

import type {method} from './assets.interface'

export const appearsTupled = (request: any): boolean =>
  isArray(request[0]) && isArray(request[0][0])

export const toWildcard = (pattern: string) => normalize(`${pattern}/**/*`)

export const assets: method = async function assets(
  ...request: Array<
    | string
    | CopyPlugin.ObjectPattern
    | Array<string | [string, string] | CopyPlugin.ObjectPattern>
  >
): Promise<Bud> {
  /**
   * tsc will complain about `this` context being lost
   * when destructuring bud even though the context of
   * this function will be bound.
   */
  const app = this as Bud

  /**
   * Replace a leading dot with the project path
   */
  const fromDotRel = (pattern: string) =>
    pattern?.startsWith('./') ? pattern.replace('./', app.path()) : pattern

  /**
   * Take an input string and return a {@link CopyPlugin.ObjectPattern}
   */
  function makePatternObject(input: string): CopyPlugin.ObjectPattern {
    return {
      from: fromDotRel(input),
      to: app.path('@name'),
      context: input.startsWith('/') ? undefined : app.path('@src'),
      noErrorOnMissing: true,
    }
  }

  /**
   * Handle tuple set
   *
   * @param tuple - [origin, destination]
   * @returns
   */
  const makeFromTo = ([from, to]: [string, string]) => ({
    from,
    to,
    noErrorOnMissing: true,
  })

  /**
   * Parse a request item
   */
  const parse = (request: string | CopyPlugin.ObjectPattern) => {
    return isString(request) ? makePatternObject(request) : request
  }

  if (appearsTupled(request)) {
    app.extensions.get('copy-webpack-plugin').setOptions(options => ({
      ...(options ?? {}),
      patterns: [
        ...(options?.patterns ?? []),
        ...request.flat().map(makeFromTo),
      ],
    }))

    return app
  }

  app.extensions.get('copy-webpack-plugin').setOptions(options => ({
    ...(options ?? {}),
    patterns: [...(options?.patterns ?? []), ...request.flat().map(parse)],
  }))

  return app
}
