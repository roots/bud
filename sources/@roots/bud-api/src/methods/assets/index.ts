import type {Bud} from '@roots/bud-framework'
import type {Plugin as CopyPlugin} from '@roots/bud-support/copy-webpack-plugin'

import {isAbsolute, join, parse, relative} from 'node:path'

type FromToTuple = [string, string]

export type Parameters = [
  (
    | Array<CopyPlugin.ObjectPattern | FromToTuple | string>
    | CopyPlugin.ObjectPattern
    | string
  ),
  Partial<CopyPlugin.ObjectPattern>?,
]

export interface assets {
  (...parameters: Parameters): Promise<Bud>
}

export const assets: assets = async function assets(
  this: Bud,
  request,
  overrides = {},
) {
  const app = this as Bud

  const makePatternObjectFromString = fromStringFactory(app, overrides)
  const makePatternObjectFromTuple = fromTupleFactory(app, overrides)
  const arrayedRequest = !Array.isArray(request) ? [request] : request

  const valueMapper = (
    item: CopyPlugin.ObjectPattern | FromToTuple | string,
  ) => {
    if (typeof item === `string`) {
      return makePatternObjectFromString(item)
    }
    if (Array.isArray(item)) {
      return makePatternObjectFromTuple(...item)
    }
    return {...item, ...overrides}
  }

  app.extensions
    .get(`@roots/bud-extensions/copy-webpack-plugin`)
    .set(`patterns`, (patterns = []) => [
      ...patterns,
      ...arrayedRequest.map(valueMapper),
    ])

  app.api.logger.success(
    `bud.copy: ${arrayedRequest.length} asset patterns added`,
  )

  return app
}

/**
 * Take an input string and return a {@link CopyPlugin.ObjectPattern}
 */
export const fromStringFactory =
  (app: Bud, overrides: Partial<CopyPlugin.ObjectPattern>) =>
  (from: string): CopyPlugin.ObjectPattern => {
    from = isAbsolute(from)
      ? relative(app.path(`@src`), from)
      : app.relPath(from)
    from = parse(from).ext?.length > 0 ? from : join(from, `**`, `*`)

    return {
      context: app.path(`@src`),
      from,
      globOptions: {dot: false},
      noErrorOnMissing: true,
      to: app.relPath(`@file`),
      ...overrides,
    }
  }

/**
 * Take an input [from,to] tuple and return a {@link CopyPlugin.ObjectPattern}
 */
export const fromTupleFactory =
  (app: Bud, overrides: Partial<CopyPlugin.ObjectPattern>) =>
  (from: string, to: string): CopyPlugin.ObjectPattern => {
    from = isAbsolute(from)
      ? relative(app.path(`@src`), from)
      : app.relPath(from)

    to = isAbsolute(to) ? relative(app.path(`@dist`), to) : app.relPath(to)
    to =
      parse(from).ext?.length > 0 || to.includes(`[`)
        ? to
        : app.relPath(to, `@file`)

    return {
      context: app.path(`@src`),
      from,
      globOptions: {dot: false},
      noErrorOnMissing: true,
      to,
      ...overrides,
    }
  }
