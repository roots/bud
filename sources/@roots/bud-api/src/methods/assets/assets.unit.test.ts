import {Bud, factory} from '@repo/test-kit/bud'

import {assets} from './assets.method'

describe('bud.assets', function () {
  let bud: Bud
  let instance: typeof assets

  beforeEach(async () => {
    bud = await factory()
    instance = assets.bind(bud)
  })

  describe('expectations', () => {
    it('should be a function', () => {
      expect(bud.assets).toBeInstanceOf(Function)
    })

    it('should have copy-webpack-plugin available', () => {
      expect(bud.extensions.has('copy-webpack-plugin')).toBeTruthy()
    })
  })

  it('should add job when passed an array of strings', async () => {
    await instance(['images'])

    expect(
      bud.extensions.get('copy-webpack-plugin').getOption('patterns'),
    ).toMatchSnapshot(
      expect.arrayContaining([
        expect.objectContaining({
          from: expect.stringContaining('images'),
          to: expect.stringContaining('dist/images/[path][name][ext]'),
          context: expect.stringContaining('src'),
          toType: 'template',
          noErrorOnMissing: true,
        }),
      ]),
    )
  })

  it('should add jobs when passed an array of tuples', async () => {
    await instance([
      [bud.path('@src/images'), bud.path('@dist/images')],
      [
        bud.path('@src/fonts/font.woff'),
        bud.path('@dist/fonts/font.woff'),
      ],
    ])

    const patterns = bud.extensions
      .get('copy-webpack-plugin')
      .getOption('patterns')
      ?.slice(0, 2)

    expect(patterns).toMatchSnapshot(
      expect.arrayContaining([
        expect.objectContaining({
          from: bud.path('@src/images'),
          to: bud.path('@dist/images'),
          context: expect.stringContaining('src'),
          toType: 'template',
          noErrorOnMissing: true,
        }),

        expect.objectContaining({
          from: bud.path('@src/fonts/font.woff'),
          to: bud.path('@dist/fonts/font.woff'),
          context: expect.stringContaining('src'),
          toType: 'template',
          noErrorOnMissing: true,
        }),
      ]),
    )
  })

  it('should add jobs when passed an object', async () => {
    const input = {
      from: bud.path('@src/images'),
      to: bud.path('@dist/images'),
    }

    await instance(input)

    const patterns = bud.extensions
      .get('copy-webpack-plugin')
      .getOption('patterns')

    expect(patterns).toMatchSnapshot(
      expect.arrayContaining([
        expect.objectContaining({
          from: bud.path('@src/images'),
          to: bud.path('@dist/images'),
        }),
      ]),
    )
  })
})
