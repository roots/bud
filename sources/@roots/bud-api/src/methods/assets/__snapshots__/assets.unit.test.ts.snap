// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bud.assets should add job when passed an array of strings 1`] = `
Array [
  Object {
    "context": "/Users/kellymears/code/git/roots/bud/tests/util/project/src",
    "from": "/Users/kellymears/code/git/roots/bud/tests/util/project/src/images",
    "noErrorOnMissing": true,
    "to": "/Users/kellymears/code/git/roots/bud/tests/util/project/dist/images/[path][name][ext]",
    "toType": "template",
  },
]
`;

exports[`bud.assets should add jobs when passed an array of tuples 1`] = `
Array [
  Object {
    "context": "/Users/kellymears/code/git/roots/bud/tests/util/project/src",
    "from": "/Users/kellymears/code/git/roots/bud/tests/util/project/src/images",
    "noErrorOnMissing": true,
    "to": "/Users/kellymears/code/git/roots/bud/tests/util/project/dist/images",
    "toType": "template",
  },
  Object {
    "context": "/Users/kellymears/code/git/roots/bud/tests/util/project/src",
    "from": "/Users/kellymears/code/git/roots/bud/tests/util/project/src/fonts/font.woff",
    "noErrorOnMissing": true,
    "to": "/Users/kellymears/code/git/roots/bud/tests/util/project/dist/fonts/font.woff",
    "toType": "template",
  },
]
`;

exports[`bud.assets should add jobs when passed an object 1`] = `
Array [
  Object {
    "from": "/Users/kellymears/code/git/roots/bud/tests/util/project/src/images",
    "to": "/Users/kellymears/code/git/roots/bud/tests/util/project/dist/images",
  },
]
`;
